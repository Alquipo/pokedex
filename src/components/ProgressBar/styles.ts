import styled, { css } from 'styled-components'
import { ProgressBarProps } from '.'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    height: 2rem;
    width: 100%;
    background-color: ${theme.colors.lightGray};
    border-radius: ${theme.border.radius.small};
    margin: 5rem;
  `}
`
type fillerStylesProps = {
  progressBar: number | string
} & Pick<ProgressBarProps, 'bgColor' | 'size'>

const fillerStylesModifiers = {
  base: (completed: string | number) => css`
    width: calc(${completed} / 2.5);
  `,

  min: (completed: string | number) => css`
    width: calc(${completed} / 4.5);
  `,

  max: (completed: string | number) => css`
    width: calc(${completed} / 6);
  `
}

export const fillerStyles = styled.div<fillerStylesProps>`
  ${({ theme, bgColor, size, progressBar }) => css`
    height: 100%;

    background-color: ${theme.colors.pokemonType[bgColor]};
    border-radius: inherit;

    transition: background-color 1.5s ease-in, width 1.5s ease-in-out;

    display: flex;
    justify-content: flex-end;
    align-items: center;

    ${!!size && fillerStylesModifiers[size](progressBar)}
  `}
`
export const labelStyles = styled.span`
  ${({ theme }) => css`
    padding-right: ${theme.spacings.xxsmall};
    color: ${theme.colors.white};
    font-weight: 500;
  `}
`
