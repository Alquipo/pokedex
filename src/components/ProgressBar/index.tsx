import { PokemonTypesProps } from 'components/PokemonCard'
import { useEffect, useState } from 'react'
import { calculateMaxStats } from 'utils/calculateStatsPokemon'
import * as S from './styles'

export type Sizes = 'base' | 'min' | 'max'

export type ProgressBarProps = {
  bgColor: PokemonTypesProps
  stats: number
  hasFunction?: boolean
  size?: Sizes
}
const ProgressBar = ({
  bgColor,
  stats,
  hasFunction,
  size = 'base'
}: ProgressBarProps) => {
  const [completedFunction, setCompletedFunction] = useState<number>(0)
  const [statusPokemon, setStatusPokemon] = useState<number>(0)

  useEffect(() => {
    switch (size) {
      case 'base': {
        setInterval(
          () => setCompletedFunction(Math.floor(Math.random() * 250) + 1),
          1500
        )
        setStatusPokemon(stats)

        break
      }

      case 'min': {
        setInterval(
          () => setCompletedFunction(Math.floor(Math.random() * 400) + 1),
          1500
        )

        const statusMinPokemon = Math.floor(
          Math.floor((2 * stats * 100) / 100 + 5) * 0.9
        )
        setStatusPokemon(Math.trunc(statusMinPokemon))

        break
      }

      case 'max': {
        setInterval(
          () => setCompletedFunction(Math.floor(Math.random() * 500) + 1),
          1500
        )

        setStatusPokemon(calculateMaxStats(stats))
        break
      }
    }
  }, [size, stats])
  return (
    <S.Wrapper>
      {hasFunction ? (
        <S.fillerStyles
          size={size}
          bgColor={bgColor}
          progressBar={`${completedFunction}%`}
        >
          <S.labelStyles>{completedFunction}</S.labelStyles>
        </S.fillerStyles>
      ) : (
        <S.fillerStyles
          size={size}
          bgColor={bgColor}
          progressBar={`${statusPokemon}%`}
        >
          <S.labelStyles>{statusPokemon}</S.labelStyles>
        </S.fillerStyles>
      )}
    </S.Wrapper>
  )
}

export default ProgressBar
