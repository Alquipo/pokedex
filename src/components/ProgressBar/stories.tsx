import { Story, Meta } from '@storybook/react/types-6-0'
import ProgressBar, { ProgressBarProps } from '.'

export default {
  title: 'ProgressBar',
  component: ProgressBar,
  args: {
    bgColor: 'Grass',
    stats: 190,
    size: 'base'
  },
  parameters: {
    backgrounds: {
      default: 'dark'
    }
  }
} as Meta

export const Default: Story<ProgressBarProps> = (args) => (
  <div
    style={{
      width: '500px',
      display: 'flex',
      flexDirection: 'column'
    }}
  >
    <ProgressBar {...args} />
  </div>
)
