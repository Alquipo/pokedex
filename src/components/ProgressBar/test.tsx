import { render, screen } from '@testing-library/react'

import ProgressBar from '.'

describe('<ProgressBar />', () => {
  it('should render the heading', () => {
    const { container } = render(<ProgressBar />)

    expect(screen.getByRole('heading', { name: /ProgressBar/i })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
