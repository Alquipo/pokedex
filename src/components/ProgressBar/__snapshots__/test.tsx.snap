// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBar /> should render the heading 1`] = `
<main
  class=""
>
  <h1>
    ProgressBar
  </h1>
</main>
`;
